src/frontend-samples/tree-samples/drag-and-drop-tree/DragAndDropTree.tsx

```diff
@@ -0,0 +1,238 @@
+    setCanDrop(newCanDrop);
+    return newCanDrop;
+  }
+
+  const elementRef = React.useRef<HTMLDivElement>(null);
+  const [dropArea, setDropArea] = React.useState<DropArea>(DropArea.Inside);
+  function handleHover(item: NodeDragObject, monitor: DropTargetMonitor): void {
+    const cursorY = monitor.getClientOffset()?.y;
+    const dropTargetY = elementRef.current?.getBoundingClientRect().y;
+    if (cursorY !== undefined && dropTargetY !== undefined) {
+      setDropArea(determineDropArea(25, 7, cursorY - dropTargetY));
+      handleCanDrop(item);
+    }
+  }
+
+  function handleDrop(item: NodeDragObject): void {
+    const { parentId, index } = getDropLocation(treeModelSource.getModel(), props.node, dropArea);
+    moveNode(treeModelSource, treeNodeLoader, item.id, parentId, index);
+  }
+
+  const [{ isHovered }, dropTargetRef] = useDrop({
+    accept: "tree-node",
+    hover: handleHover,
+    canDrop: handleCanDrop,
+    drop: handleDrop,
+    collect: (monitor) => ({ isHovered: monitor.isOver() }),
+  });
+
+  const isDropAreaDisplayed = isHovered && canDrop;
+  const nodeStyle: React.CSSProperties = {
+    height: 25,
+    ...(isDropAreaDisplayed && dropArea === DropArea.Inside && { background: "var(--buic-row-hover)" }),
+  };
+  return (
+    <div ref={mergeRefs([dragSourceRef, dropTargetRef, elementRef])}>
+      <div style={{ height: 0 }}>
+        {isDropAreaDisplayed && dropArea === DropArea.Above && <NodeInsertMarker topOffset={0} />}
+        {isDropAreaDisplayed && dropArea === DropArea.Below && <NodeInsertMarker topOffset={25} />}
+      </div>
+      <SimpleTreeNode style={nodeStyle} isHoverDisabled={isDragging} {...props} />
+    </div>
+  );
+};
+
+function determineDropArea(areaHeight: number, edgeTargetHeight: number, dropOffset: number): DropArea {
+  if (dropOffset < edgeTargetHeight) {
+    return DropArea.Above;
+  }
+
+  if (dropOffset > (areaHeight - edgeTargetHeight)) {
+    return DropArea.Below;
+  }
+
+  return DropArea.Inside;
+}
+
+interface NodeDragObject {
+  type: "tree-node";
+  id: string;
+}
+
+interface DropLocation {
+  parentId: string | undefined;
+  index: number;
+}
+
+function getDropLocation(treeModel: TreeModel, hoveredNode: TreeModelNode, dropArea: DropArea): DropLocation {
+  switch (dropArea) {
+    case DropArea.Above:
+      return {
+        parentId: hoveredNode.parentId,
+        index: (treeModel as MutableTreeModel).getChildOffset(hoveredNode.parentId, hoveredNode.id)!,
+      };
+
+    case DropArea.Inside:
+      return { parentId: hoveredNode.id, index: 0 };
+
+    case DropArea.Below:
+      if (hoveredNode.isExpanded) {
+        return {
+          parentId: hoveredNode.id,
+          index: 0,
+        };
+      }
+
+      return {
+        parentId: hoveredNode.parentId,
+        index: (treeModel as MutableTreeModel).getChildOffset(hoveredNode.parentId, hoveredNode.id)! + 1,
+      };
+  }
+}
```

Comment:
What's your opinion on moving `getDropLocation` and `determineDropArea` to imodeljs?

Reply:
I'd be strongly against making them a part of imodeljs libraries because these methods are highly specialized for this particular interaction design. You could imagine some use cases where only reordering or reparenting is needed, which would require a whole lot different visual and interaction design. Besides, these methods as they are currently are not tough to figure out from scratch.

Reply:
Agreed. Thanks for explanation.